modules/fedimint-mint-client/src/backup/recovery.rs

```diff
@@ -271,20 +284,32 @@ impl MintRestoreInProgressState {
     /// errors via `sender` itself.
     ///
     /// TODO: could be internal to recovery_loop?
-    async fn await_block<'a>(
+    fn fetch_block_stream<'a>(
         api: DynGlobalApi,
         decoders: ModuleDecoderRegistry,
-        index: u64,
-    ) -> SessionOutcome {
-        loop {
-            info!(target: LOG_CLIENT_RECOVERY_MINT, index, "Awaiting block {index}");
-            match api.await_block(index, &decoders).await {
-                Ok(block) => return block,
-                Err(e) => {
-                    info!(e = %e, index, "Error trying to fetch signed block");
-                }
-            }
-        }
+        epoch_range: ops::Range<u64>,
+    ) -> impl futures::Stream<Item = (u64, SessionOutcome)> + 'a {
+        futures::stream::iter(epoch_range)
+            .map(move |block_idx| {
+                let api = api.clone();
+                let decoders = decoders.clone();
+                Box::pin(async move {
+                    info!(block_idx, "Fetching epoch");
+
+                    let block = loop {
+                        info!(target: LOG_CLIENT_RECOVERY_MINT, block_idx, "Awaiting signed block");
+                        match api.await_block(block_idx, &decoders).await {
+                            Ok(block) => break block,
+                            Err(e) => {
+                                info!(e = %e, block_idx, "Error trying to fetch signed block");
+                            }
+                        }
+                    };
+
+                    (block_idx, block)
+                })
+            })
+            .buffered(8)
```

Comment:
maybe bigger number

Reply:
8 isn't super bad, I think it should be the number of downloads that can complete in one round trip time to not waste any time waiting. For active federations with medium client/server latency this would actually be a ok number, but since we currently mostly have inactive ones we could increase it by an order of magnitude.

Reply:
It's important to remember from the queue processing theory that increasing queue times is usually harmful: wastes resources and increases latencies. 

On top of my head an example: https://en.wikipedia.org/wiki/Bufferbloat

The role of a buffer is mostly to amortize the variations in the processing/job arrival times (burstiness, etc) on both ends, and allow both sides to work in parallel.

[`buffered` also polls buffered futures in parallel](https://docs.rs/futures-util/0.3.30/src/futures_util/stream/stream/buffered.rs.html#63) effectively increasing maximum parallelism level - I guess I could imagine implementation that polls only the first one, while buffering up to N resolved ones or something. Anyway, parallism is welcomed speedup here as well, but increasing it too much might not yield much improvement and considerably increase the load on the Federation, which is not great either.

So my feeling is that an 8 is a solid choice. :D

@Maan2003 @elsirion